### ISSUE TITLE
Solution design for a new Greenshot file format

### ISSUE THREAD

**Issue by Lakritzator** [greenshot/greenshot#375]
**Is your feature request related to a problem? Please describe.**
The current Greenshot uses as it's own file-format a PNG, followed by the serialized surface (the canvas) elements. Unfortunately this is not only very proprietary, it's also not stable over versions. This means that something written in Greenshot 1.2 cannot be loaded in 1.3 etc.

**Describe the solution you'd like**
I'd like a file format which is clearly documented, and even could be used by other applications.

**Describe alternatives you've considered**
Use SVG

**Additional context**
...

**Comment by Lakritzator** [comment]
I've extracted most requirements of a potential file format from the previous information.
So we need something which can store a background image, and draw different annotations on top of it.
These annotations could be different shapes (lines, arrows, circles, rectangles, freehand) in different colors, text and images.

It's important to be able to create this file from the information in the editor, and be able to read the file to build the editor state back again.

This means we need to store:

- The captured image
- An ordered list of elements to draw on the captured image
- The images for the elements
- The text for the elements, with the font information
- The locations of the elements
- The extra positions of the elements (e.g. the tail of the speech bubble)
- The points (path) for the freehand drawing
- For the shape elements the line thickness, line color, fill color, arrows 

Some additional requirements:

- Multiple layers
- Lossless (no quality loss)
- All image elements are stored with their original information, so they can be resized later without additional quality loss.
- Meta data (like timestamp etc)
- Line caps & dash (to enable more ways to draw elements)


Some optional requirements:

- The file can be read by other image viewers, by piggy-bagging on an existing format. e.g. store all the information in a new PNG chunk.

**Comment by Lakritzator** [comment]
I guess it should be possible to use SVG, but the specs are very extensive and complex. Also the goal of SVG is not the same as we need for Greenshot.
I also looked at draw.io, but my first impression was that it's very complex and I'm not sure if it wouldn't cost me more time to use it and maintain it then to make my own.

I wrote a [tweet](https://twitter.com/lakritzator/status/1564312166668402689?s=20&t=vRMv5-0uCopAUFDXtEa8AQ), to see if someone has an idea for this.

Otherwise I could **imagine** to build something like this:

- Build a set of classes to represent a json format (due to the ease of de/serializing with .NET) for the information I need.
- Convert the in memory representation of the editor information to this set of classes, and serialize this to a json format
- Use a .zip archive to store the files like the capture, the serialized json and all other images.
- Generate the current output (result) as .PNG and add a custom chunk with the .ZIP as the stream. (Extend [ImageSharp](https://github.com/SixLabors/ImageSharp/issues/251) for this?)

**Comment by Lakritzator** [comment]
I just marked it as a blocker, maybe this is not everybody's opinion, but there seem to be more people depending on this feature as shows and building something without backwards compatibility is really not going to make them happy.

**Comment by Vasilich** [comment]
Yes, for our documentation chain we keep screenshots with drawings from Greenshot, saving it in .greenshot format. 
My colleagues were happy to see zoom function coming with unstable releases, but later on found out that new (unstable 1.3) version cannot read greenshot drawings created with stable (1.2) version. Right now we use 2 versions of Greenshot installed in different folders, and start old one if we need to access older pictures. But this is surely inconvenient.
SVG can keep most of containers and its properties, except start counter and filters like pixelation size, but this can be achieved by using custom attributes - it is just a question whether you want to make resulting SVG really SVG-compliant so that it should be editable by real SVG editors (like inkscape)? Though SVG format can contain a lot of things in different notations (like CSS support), it doesn't have to be done - we can just use that parts of SVG, that we need.

**Comment by Lakritzator** [comment]
@Vasilich Thank you for the feedback, I'm not sure how to go forward though.
Sticking to a standard, also means we need to "maintain" it, and I am not sure about the added bennefits or even if it can be made to work. Let's see...

About backwards compatibility, I will try to see if I can make a tool to export 1.2 in such a way that 1.3 can read it. Having an additional .exe might increase the size, but I do not see many other ways. We might also just deliver this as an additional download... don't know yet.
